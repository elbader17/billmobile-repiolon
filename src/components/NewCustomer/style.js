import { StyleSheet } from 'react-native';
import { COLORS } from '../../constants/colors';
import { FONTS } from '../../constants/fonts';
import { METRICS } from '../../constants/metrics';

const styles = StyleSheet.create({
  container: {
    height: METRICS.screenHeight,
    backgroundColor: COLORS.blue,
    alignItems: 'center',
    paddingHorizontal: 30,
    paddingTop: 15,
  },
  textBoxBtnHolder: {
    position: 'relative',
    alignSelf: 'stretch',
    justifyContent: 'center',
    marginTop: 0,
    backgroundColor: COLORS.white,
    borderRadius: 4,
  },
  picker: {
    height: 45,
    color: COLORS.grayDark
  },
  textBox: {
    fontFamily: FONTS.latoRegular,
    fontSize: FONTS.size16,
    alignSelf: 'stretch',
    height: 45,
    paddingRight: 45,
    paddingLeft: 8,
    borderWidth: 1,
    paddingVertical: 0,
    borderColor: COLORS.gray,
    borderRadius: 2,
  },
  submit: {
    backgroundColor: COLORS.red,
    height: 40,
    borderRadius: 2,
    width: 285,
  },
  submitDisabled: {
    backgroundColor: COLORS.gray,
    width: 285,
    height: 40,
    borderRadius: 2,
  },
  submitText: {
    textAlign: 'center',
    fontFamily: FONTS.latoRegular,
    fontSize: FONTS.size14,
    color: COLORS.white,
  },
  textRegular14White: {
    fontFamily: FONTS.latoRegular,
    fontSize: FONTS.size14,
    color: COLORS.white,
    paddingVertical: 10,
    textAlign: 'center',
  },
  textRegular11White: {
    fontFamily: FONTS.latoRegular,
    fontSize: FONTS.size11,
    color: COLORS.white,
    paddingVertical: 5,
  },
  red: {
    color: COLORS.red,
  },
  inLine: {
    flexDirection: "row",
    position: 'absolute',
    bottom:80
  },
  textFooterA: {
    fontFamily: FONTS.latoRegular,
    fontSize: FONTS.size11,
    color: COLORS.grayDark,
    marginTop: 110,
  },
  textFooterB: {
    fontFamily: FONTS.latoRegular,
    fontSize: FONTS.size11,
    color: COLORS.grayDark,
    fontWeight: 'bold',
  },
  headerText: {
    fontFamily: FONTS.latoRegular,
    fontSize: FONTS.size16,
    fontWeight: 'normal',
    color: COLORS.blue,
  },
  lineWhite: {
    backgroundColor: COLORS.white,
    height: 1,
    width: '100%',
    marginTop: 20,
    marginBottom: 10,
  },
});

export default styles;
