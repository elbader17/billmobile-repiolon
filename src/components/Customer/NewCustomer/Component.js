import React from 'react';
import { View, Text, Picker, ScrollView, TouchableOpacity } from 'react-native';
import { TextField } from 'react-native-material-textfield';
import { Button } from "react-native-elements";
import LinearGradient from 'react-native-linear-gradient';
import { COLORS, GRADIANTBLUE2, COLORGBL } from '../../../constants/colors';
import { CONDITION_IVA } from '../../../constants/fiscal_identity';
import  { iconMenuBack }  from '../../../constants/icons';
import { validateCuit } from '../../../utils/identity';
import { XY } from '../../../constants/gradientCoord';
import style from '../style';

class NewCustomer extends React.Component {

  constructor(props) {
    super(props);
    const customer = this.props.navigation.getParam('customer', this.defaultCustomer());

    this.state = {
      name: customer.attributes.name,
      category: customer.attributes.category,
      identification: customer.attributes.identification,
      customerId: customer.id,
      loading: false,
      error: undefined
    }
  }

  defaultCustomer= () => {
    return {
      attributes: {
        name: '',
        identification: '',
        category: 'monotributo',
      },
    };
  }

  static navigationOptions = ({navigation}) => {
    return {
      title: 'Nuevo Cliente',
      headerTransparent: true,
      headerStyle: { elevation: 0 },
      headerTitleStyle: style.headerText,
      headerTintColor: COLORS.white,
      headerLeft:( 
        <TouchableOpacity onPress={()=> navigation.goBack()}>
          {iconMenuBack}
        </TouchableOpacity>
      ) 
    }  
  }

  saveCustomer = () => {
    const { name, category, identification, customerId } = this.state;
    if (validateCuit(identification)) {
      this.setLoading(true);
      const { saveFiscalIdentity, navigation, type } = this.props;
      saveFiscalIdentity(name, identification, category, customerId)
        .then(() => {
          type === 'collection' ? navigation.navigate('CustomersList'): navigation.navigate('Invoice');
          this.setLoading(false);
        })
    } 
    else this.setState({error: 'CUIT Inválido'})
  }

  setName = (value) => this.setState({ name: value})
  setIdentification = (value) => this.setState({ identification: value })
  setCategory = (value) => this.setState({ category: value })
  setLoading = (bool) => this.setState({ loading: bool })

  render() {
    return(
      <LinearGradient 
        colors={ GRADIANTBLUE2 } 
        style={ style.container }
        start={ XY.startV } 
        end={ XY.endV }>

          <View style={style.containerBody}>
            <ScrollView>
              <View style={style.containerInputs}>
                
                <View style={style.inputPicker}>
                  <Picker
                    selectedValue={this.state.category}
                    style= {style.picker}
                    onValueChange={this.setCategory}>
                      { CONDITION_IVA.map((i, index) => (
                        <Picker.Item 
                          key={index} 
                          color='gray' 
                          label={i.label} 
                          value={i.value}/>
                       ))}
                  </Picker>
                </View>

                <Text style={[style.textRegular12White, {marginTop: 5}]}>
                  Condición frente al IVA
                </Text>

                <TextField
                  title='Su información fiscal se cargará con su CUIT'
                  label='Número de CUIT'
                  value={this.state.identification}
                  onChangeText={this.setIdentification}
                  onFocus={()=>{this.setState({error: undefined})}}
                  keyboardType='numeric'
                  tintColor={COLORS.blueLight}
                  textColor= {COLORS.gray}
                  baseColor={COLORS.white}
                  lineWidth={1}
                  labelFontSize={15}
                  labelPadding={6}
                  characterRestriction={11}
                  error={this.state.error}
                  errorColor={'#ff9999'}
                />
                
                <TextField
                  title='Opcional'
                  label='Nombre de Fantasía'
                  value={this.state.name}
                  onChangeText={this.setName}
                  textColor= {COLORS.gray}
                  tintColor={COLORS.blueLight}
                  baseColor={COLORS.white}
                  lineWidth={1}
                  labelFontSize={15}
                  labelPadding={6}
                />
              </View>
            </ScrollView>
          </View>

          <View style={[style.containerFooter, {alignItems: 'center'}]}>
            <Button
              title={this.props.type==='invoice' ? 'Añadir al Comprobante' : 'Guardar'}
              TouchableComponent={TouchableOpacity}
              onPress={this.saveCustomer}
              buttonStyle={ style.buttonSave }
              titleStyle={ style.textRegular16White }
              loading={this.state.loading}
              ViewComponent={LinearGradient}
              linearGradientProps={COLORGBL}
            /> 
          </View>

      </LinearGradient>
    )
  }
}

export default NewCustomer;