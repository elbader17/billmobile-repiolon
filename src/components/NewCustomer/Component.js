import React from 'react';
import { View, Text, Alert, Picker, TextInput, ScrollView, KeyboardAvoidingView } from 'react-native';
import { Button } from "react-native-elements";
import { withNavigation } from 'react-navigation';
import  { validateCuit } from '../../utils/identity';
import { METRICS } from '../../constants/metrics';
import { COLORS } from '../../constants/colors';
import style from './style';
import {
  CONDITION_IVA,
} from '../../constants/fiscal_identity';

class NewCustomer extends React.Component {

  constructor(props) {
    super(props);
    const { fiscalIdentity } = this.props;
    this.state = {
      conditionIva: '',
      category: fiscalIdentity.category,
      identification: fiscalIdentity.identification,
      name: fiscalIdentity.name,
      loading: false
    }
  }

  newCustomer = () => {
    const { name, category, identification } = this.state;
    const {
      addFiscalIdentity,
      fiscalIdentity,
      navigation,
    } = this.props;
    this.setLoading(true);
    addFiscalIdentity(name, identification, fiscalIdentity.id, navigation)
      .then(() => {
        Alert.alert("Cliente Cargado: "+this.props.fiscalIdentity.name+" "+this.props.fiscalIdentity.identification);
      })
      .catch(err => {
        Alert.alert("Error al Ingresar: ",err.message);
        this.setLoading(false);
      });
  }

  validateData = () => {
    const { identification } = this.state;
    return validateCuit(identification);
  }

  static navigationOptions = {
    title: 'NUEVO CLIENTE',
    headerTitleStyle: style.headerText,
    headerTintColor: '#3687D1',
  };

  setName = (value) => this.setState({ name: value})
  setIdentification = (value) => this.setState({ identification: value })
  setLoading = (bool) => this.setState({ loading: bool })

  render() {
    return(
      <KeyboardAvoidingView
        behavior={'padding'}
        style={{flex: 1}}
        keyboardVerticalOffset={METRICS.heightHeader}>
      <ScrollView>
        <View style={style.container}>
          <View style={style.containerInputs}>
            <Text style={style.textRegular14White}>CONDICIÓN FRENTE AL IVA</Text>
            <View style={style.boxBtnHolder}>
              <Picker
                selectedValue={this.state.conditionIva}
                style= {style.picker}
                onValueChange={itemValue => this.setState({ conditionIva: itemValue })}>
                  {CONDITION_IVA.map((i, index) => (
                    <Picker.Item key={index} color='gray' label={i.label} value={i.value} />
                ))}
              </Picker>
            </View>
            <Text style={style.textRegular14White}>NÚMERO DE CUIT</Text>
            <View style={ style.boxBtnHolder }>
              <TextInput
                onChangeText={this.setIdentification}
                placeholder="00-00000000-0"
                value={this.state.identification}
                placeholderTextColor={COLORS.grayLight}
                value={this.state.identification}
                style={[style.textRegular16GrayDark,style.marginLeft5]}
                keyboardType='numeric'
              />
            </View>
          <Text style={ style.textRegular12White }>La información fiscal de los clientes se cargan automaticamente poniendo su N° de CUIT</Text>

          <View style={style.lineWhite}></View>

          <Text style={style.textRegular14White} >NOMBRE DE FANTASÍA (OPCIONAL)</Text>
          <View style={ style.boxBtnHolder }>
            <TextInput
              style={[style.textRegular16GrayDark,style.marginLeft5]}
              onChangeText={this.setName}
              value={this.state.name}
              placeholder="Inserta el Nombre"
              value={this.state.name}
              placeholderTextColor={COLORS.grayLight}
            />
          </View>
          </View>
      </View>
      </ScrollView>
      <Button
        onPress={this.newCustomer}
        title='GUARDAR'
        buttonStyle={ style.buttonSave }
        titleStyle={ style.textRegular14WhiteBold }
        disabledStyle= { style.buttonSaveDisabled }
        disabledTitleStyle = { style.textRegular14WhiteBold }
        disabled={!this.validateData() }
        loading = {this.state.loading}
      />
      </KeyboardAvoidingView>
    )
  }
}


export default withNavigation(NewCustomer);
