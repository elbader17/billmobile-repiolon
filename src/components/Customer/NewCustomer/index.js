import { connect } from 'react-redux';
import NewCustomer from './Component';
import { createCustomer } from '../../../app/customers/action';

const mapStateToProps = state => ({
  jwtToken: state.authentication.jwtToken,
  fiscalIdentity: state.fiscalIdentity,
});

const mapDispatchToProps = (dispatch) => {
  return {
    saveFiscalIdentity: (name, identification, category, id, navigation) => {
      dispatch(createCustomer({ name, identification, category }))
        .then(() => {
          navigation.navigate('CustomerList')
        });
    },
    type: 'collection'
  };
};

const component = connect(
  mapStateToProps,
  mapDispatchToProps,
)(NewCustomer);

export default component;