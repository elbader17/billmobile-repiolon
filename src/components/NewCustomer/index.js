/* eslint-disable import/no-named-as-default */
import { connect } from 'react-redux';
import NewCustomer from './Component';
import { addFiscalIdentityToInvoice } from '../../app/fiscal_identity/actions';

const mapStateToProps = state => ({
  jwtToken: state.authentication.jwtToken,
  fiscalIdentity: state.fiscalIdentity,
});

const mapDispatchToProps = dispatch => {
  return {
    addFiscalIdentity: (name, cuit, navigation) => {
      dispatch(addFiscalIdentityToInvoice(name, cuit))
        .then(() => navigation.navigate('Invoice'));
    },
  };
};

const component = connect(
  mapStateToProps,
  mapDispatchToProps,
)(NewCustomer);

export default component;
