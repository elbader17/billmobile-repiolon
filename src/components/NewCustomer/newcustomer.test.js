import React from 'react';
import { shallow } from 'enzyme';
import NewCustomer  from './Component';

jest.mock('../../constants/metrics', () => ({
  METRICS: {
    heightHeader: {} 
  },
}));

const promise = Promise.resolve(0);
const addFiscalIdentityMock = jest.fn(() => promise);

const setup = () => {
  const props = {
    fiscalIdentity: {
      category: '',
      identification: '',
      name: ''
    },
    addFiscalIdentity: addFiscalIdentityMock
  }
  const component = shallow(<NewCustomer {...props}/>);
  return {props, component};
};

test('render correctly component NewCustomer', () => {
  const {component} = setup();
  expect(component).toMatchSnapshot();
});

test('the property loading is false at the start', () => {
  const {component} = setup();
  expect(component.state().loading).toBe(false);
});

test('call the newCustomer function', () => {
  const {component} = setup();
  const button = component.findWhere(node => node.prop('testID') === 'buttonNewCustomer');
  button.simulate('press');
  expect(addFiscalIdentityMock).toBeCalled();
});

describe('button GUARDAR test', () => {
  
  test('button initially disabled', () => {
    const {component} = setup();
    const button = component.findWhere(node => node.prop('testID') === 'buttonNewCustomer');
    button.simulate('press');
    expect(button.props().disabled).toBe(true);
  });
      
  test('button enabled with the correct data', () => {
    const {component} = setup();
    component.setState({ 
      identification: '20363095721',
    });
    const button = component.findWhere(node => node.prop('testID') === 'buttonNewCustomer');
    expect(button.props().disabled).toBe(false);
  });
      
  test('button disabled with cuit incorrect', () => {
    const {component} = setup();
    component.setState({ 
      identification: '0363095721',
    });
    const button = component.findWhere(node => node.prop('testID') === 'buttonNewCustomer');
    expect(button.props().disabled).toBe(true);
  });
       
  test('button loading after press', () => {
    const {component} = setup();
    const button = component.findWhere(node => node.prop('testID') === 'buttonNewCustomer');
    button.simulate('press');
    expect(component.instance().state.loading).toBe(true);
  });
  
});

test('button loading after press', () => {
  const {component} = setup();
  const button = component.findWhere(node => node.prop('testID') === 'buttonNewCustomer');
  button.simulate('press');
  console.log(component.instance().state.then);
});
  