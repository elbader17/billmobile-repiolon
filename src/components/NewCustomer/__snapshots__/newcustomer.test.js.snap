// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly component NewCustomer 1`] = `
<KeyboardAvoidingView
  behavior="padding"
  enabled={true}
  keyboardVerticalOffset={Object {}}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <ScrollViewMock>
    <View
      style={
        Object {
          "backgroundColor": "#3687d1",
          "height": NaN,
          "padding": 18,
        }
      }
    >
      <View
        style={
          Object {
            "marginHorizontal": 2,
            "marginVertical": 5,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#fff",
              "fontFamily": "Lato-Regular",
              "fontSize": 14,
            }
          }
        >
          CONDICIÓN FRENTE AL IVA
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "#fff",
              "borderRadius": 4,
              "height": 93.5,
              "marginVertical": 8,
            }
          }
        >
          <Picker
            mode="dialog"
            onValueChange={[Function]}
            selectedValue=""
            style={
              Object {
                "color": "#858585",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <PickerItem
              color="gray"
              key="0"
              label="RESPONSABLE INSCRIPTO"
              value="ri"
            />
            <PickerItem
              color="gray"
              key="1"
              label="IVA EXENTO"
              value="ie"
            />
            <PickerItem
              color="gray"
              key="2"
              label="MONOTRIBUTISTA"
              value="m"
            />
            <PickerItem
              color="gray"
              key="3"
              label="RESPONSABLE NO INSCRIPTO"
              value="rni"
            />
          </Picker>
        </View>
        <Text
          style={
            Object {
              "color": "#fff",
              "fontFamily": "Lato-Regular",
              "fontSize": 14,
            }
          }
        >
          NÚMERO DE CUIT
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "#fff",
              "borderRadius": 4,
              "height": 93.5,
              "marginVertical": 8,
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            keyboardType="numeric"
            onChangeText={[Function]}
            placeholder="00-00000000-0"
            placeholderTextColor="#EEEEEE"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "color": "#858585",
                  "fontFamily": "Lato-Regular",
                  "fontSize": 15,
                },
                Object {
                  "marginLeft": 5,
                },
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <Text
          style={
            Object {
              "color": "#fff",
              "fontFamily": "Lato-Regular",
              "fontSize": 12,
            }
          }
        >
          La información fiscal de los clientes se cargan automaticamente poniendo su N° de CUIT
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "#fff",
              "height": 1,
              "marginBottom": 10,
              "marginTop": 10,
              "width": "100%",
            }
          }
        />
        <Text
          style={
            Object {
              "color": "#fff",
              "fontFamily": "Lato-Regular",
              "fontSize": 14,
            }
          }
        >
          NOMBRE DE FANTASÍA (OPCIONAL)
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "#fff",
              "borderRadius": 4,
              "height": 93.5,
              "marginVertical": 8,
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            placeholder="Inserta el Nombre"
            placeholderTextColor="#EEEEEE"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "color": "#858585",
                  "fontFamily": "Lato-Regular",
                  "fontSize": 15,
                },
                Object {
                  "marginLeft": 5,
                },
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
      </View>
    </View>
  </ScrollViewMock>
  <ForwardRef(Themed.Button)
    buttonStyle={
      Object {
        "backgroundColor": "#EE6123",
        "borderRadius": 0,
        "height": 106.5,
      }
    }
    disabled={true}
    disabledStyle={
      Object {
        "backgroundColor": "#C4C4C4",
        "borderRadius": 0,
        "height": 106.5,
      }
    }
    disabledTitleStyle={
      Object {
        "color": "#fff",
        "fontFamily": "Lato-Semibold",
        "fontSize": 14,
      }
    }
    loading={false}
    onPress={[Function]}
    testID="buttonNewCustomer"
    title="GUARDAR"
    titleStyle={
      Object {
        "color": "#fff",
        "fontFamily": "Lato-Semibold",
        "fontSize": 14,
      }
    }
  />
</KeyboardAvoidingView>
`;
